EIPS/eip-3779.md

```diff
@@ -0,0 +1,208 @@
+* `valid_jumpdest()` tests whether a destination is a JUMPDEST byte and not in immediate data.
+* `add_items()` and `remove_items()` push or pop the `const_stack` items for an instruction,
+    and return its effect on the `stack pointer`.
+    `add_items()` will `PUSH` the value of constants – all other stack items are zeroed.
+```
+    var bytecode [code_len]byte
+    var stack_depth [code_len]unsigned
+    var const_stack[1024]unsigned
+    var SP := 0
+
+    func validate(PC :=0, stack_depth:=0) boolean {
+
+        // valid instructions only
+        instruction := bytecode[PC]
+        if !valid_instruction(instruction) {
+            return false;
+        }
+
+        // if stack depth for `PC` is non-zero we have been here before 
+        // return true to break cycle in control flow graph
+        if stack_depth[PC] != 0 { 
+            return true
+        }
+        stack_depth[PC] = stack_depth
+        entrySP := SP
+
+        if instruction == JUMP {
+
+            // check for valid destination
+            dest = stack[SP]
+                        if valid!_jumpdest(dest) {
+                return false
+            }
+
+            // reset PC to destination of jump 
+            PC += jumpdest
+                        --SP
+            continue
+        }
+        if instruction == JUMPI {
+
+            // check for valid destination
+            jumpdest = stack[SP]
+            if bytecode[jumpdest] != JUMPDEST {
+                return false
+            }
+
+            // reset PC to destination of jump 
+            PC += jumpdest
+
+            // check for stack underflow and
+            // recurse to jump to code to validate
+            stack_depth := SP - entry_SP
+            if stack_depth < 0 { 
+                return false
+            }
+            --SP
+            if !validate(stack[SP]), stack_depth {
+                return false
+            }
+
+            continue 
+        }
+ 
+        // check effect of instruction on stack
+        SP -= remove_items(instruction)
+        SP += add_items(instruction)
+        if SP < 0 || 1024 < SP {
+            return false
+        }
+
+        // successful validation of path
+        if PC > code_len
+               || instruction == STOP
+               || instruction == RETURN
+               || instruction == SUICIDE {
+            return true
+        }
+
+        // advance PC according to instruction
+        PC = advance_pc(PC, instruction)
+    }
+```
+
+## Rationale
+
+The alternative to checking validity at creation time is checking it at runtime.  This hurts performance and is a DoS vulnerability.  Thus the above rules and accompanying one-pass validation algorithm.
+
+_Rule 1_ – requiring constant destinations for `JUMP` and `JUMPI` – forbids dynamic jumps.  This rule is already required by `EIP-3540`.  Jump destinations are currently checked at runtime, but static jumps can be validated at creation time.   
```

Comment:
How is this required by 3540?

Reply:
It reads "absolite" , and I misread that for "constant".  I will remove  the reference.
> JUMP/JUMPI uses an absolute offset within the code.

https://github.com/ethereum/EIPs/blob/f4f250c211a5d9628ec81c47d15de45107d6acb3/EIPS/eip-3540.md#changes-to-execution-semantics